File: solutions/python/sliding-window/find-all-anagrams-in-a-string.py
# LeetCode 438 — Find All Anagrams in a String (Medium)
# Category: Sliding Window · Approach: Sliding Window
# Time: O(n) | Space: O(1)
# Source: https://leetcode.com/problems/find-all-anagrams-in-a-string/

def find_anagrams(s: str, p: str) -> list[int]:
    if len(s) < len(p):
        return []
    need = [0] * 26
    win = [0] * 26
    for ch in p:
        need[ord(ch) - 97] += 1
    res = []
    for i, ch in enumerate(s):
        win[ord(ch) - 97] += 1
        if i >= len(p):
            win[ord(s[i - len(p)]) - 97] -= 1
        if i >= len(p) - 1 and win == need:
            res.append(i - len(p) + 1)
    return res

File: solutions/typescript/sliding-window/find-all-anagrams-in-a-string.ts
// LeetCode 438 — Find All Anagrams in a String (Medium)
// Category: Sliding Window · Approach: Sliding Window
// Time: O(n) | Space: O(1)
// Source: https://leetcode.com/problems/find-all-anagrams-in-a-string/

function findAnagrams(s: string, p: string): number[] {
  const res: number[] = [];
  if (s.length < p.length) return res;
  const need = new Array(26).fill(0);
  const win = new Array(26).fill(0);
  const a = "a".charCodeAt(0);
  for (const ch of p) need[ch.charCodeAt(0) - a]++;
  for (let i = 0; i < s.length; i++) {
    win[s.charCodeAt(i) - a]++;
    if (i >= p.length) win[s.charCodeAt(i - p.length) - a]--;
    if (i >= p.length - 1) {
      let match = true;
      for (let j = 0; j < 26; j++) {
        if (win[j] !== need[j]) { match = false; break; }
      }
      if (match) res.push(i - p.length + 1);
    }
  }
  return res;
}
